Sources/OpenAPIURLSession/URLSessionTransport.swift

```diff
@@ -73,8 +73,23 @@ public struct URLSessionTransport: ClientTransport {
 
         /// Creates a new configuration with the provided session.
         /// - Parameter session: The URLSession used for performing HTTP operations.
-        ///     If none is provided, the system uses the shared URLSession.
-        public init(session: URLSession = .shared) { self.session = session }
+        ///   If none is provided, the system uses the shared URLSession.
+        public init(session: URLSession = .shared) { self.init(session: session, implementation: .platformDefault) }
+
+        enum Implementation {
+            case buffering
+            case streaming(requestBodyStreamBufferSize: Int, responseBodyStreamWatermarks: (low: Int, high: Int))
```

Comment:
My preference is to not expose `requestBodyStreamBufferSize`. It is an extremely confusing parameter that does not reflect buffering done in CFNetwork and Network framework. We could have gotten away with not using bounded stream pair if NSStream subclassing were easier in Swift, and it will likely be unnecessary in the future.

We'd like to retain the ability to fine tune these values on our platforms without needing developers to change their code, e.g. reducing the buffer size on low memory devices like Apple Watch, or even adjusting it dynamically based on memory pressure / app state. Most developers are not equipped to decide the best buffer size to use.

Would they be better as environment variables for advanced use cases that we don't need to commit as APIs?

Reply:
> My preference is to not expose `requestBodyStreamBufferSize`. It is an extremely confusing parameter that does not reflect buffering done in CFNetwork and Network framework. 
> ...
> We'd like to retain the ability to fine tune these values on our platforms without needing developers to change their code, e.g. reducing the buffer size on low memory devices like Apple Watch, or even adjusting it dynamically based on memory pressure / app state. Most developers are not equipped to decide the best buffer size to use.

This is _not_ exposed. This `enum Implementation` is an internal type, only used in the `internal init(session:implementation:)`. The only public initializer is `public init(session:)` which calls `internal init(session:implementation:)` with `Implementation.platformDefault`, which has some hard-coded buffer sizes and is also internal, which could allow us to tweak on a per-platform basis.

> We could have gotten away with not using bounded stream pair if NSStream subclassing were easier in Swift, and it will likely be unnecessary in the future.

Oh, that's interesting to know. I'll watch this space 👀 

> Would they be better as environment variables for advanced use cases that we don't need to commit as APIs?

Hadn't considered an environment variable based approach. I think we could incorporate that too, for advanced users.

One thing to call out here that we were _considering_ exposing was a way to _force_ buffering. Right now, when providing an `HTTPBody` with `iterationBehavior == .single`, the observed behaviour is different when faced with a HTTP redirect. Specifically, it will succeed when using buffering, but _fail_ when using streaming, because we cannot rewind the request body stream.

We have other ideas on how we could solve this—e.g. wrapping the provided `AsyncSequence` in another that buffers the first N bytes and might have a better chance of replaying on redirect—but in general we may want to expose the ability to just turn off streaming.

Curious if you have any thoughts on this?

//cc @dnadoba @czechboy0

Reply:
Maybe let's start with the parameter being internal, and only make it public if a genuine use case comes up.

I'm not a fan of libraries checking environment variables, it makes them difficult to test and reason about.